flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClientEndpoint.java

```diff
@@ -0,0 +1,263 @@
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.DefaultFullHttpRequest;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.FullHttpRequest;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.FullHttpResponse;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpClientCodec;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpHeaders;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpObjectAggregator;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpVersion;
+import org.apache.flink.shaded.netty4.io.netty.handler.ssl.SslHandler;
+
+import com.fasterxml.jackson.core.JsonParseException;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonMappingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.net.ssl.SSLEngine;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.StringWriter;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+
+/**
+ * This client is the counter-part to the {@link RestServerEndpoint}.
+ */
+public class RestClientEndpoint {
+	private static final Logger LOG = LoggerFactory.getLogger(RestClientEndpoint.class);
+
+	private static final ObjectMapper objectMapper = RestMapperUtils.getStrictObjectMapper();
+
+	private final String configuredTargetAddress;
+	private final int configuredTargetPort;
+	private final SSLEngine sslEngine;
+
+	private Bootstrap bootstrap;
+
+	private final ClientHandler handler = new ClientHandler();
+
+	private CompletableFuture<?> lastFuture = CompletableFuture.completedFuture(null);
+
+	private final Executor directExecutor = Executors.directExecutor();
+
+	public RestClientEndpoint(RestClientEndpointConfiguration configuration) {
+		this.configuredTargetAddress = configuration.getTargetRestEndpointAddress();
+		this.configuredTargetPort = configuration.getTargetRestEndpointPort();
+		this.sslEngine = configuration.getSslEngine();
+	}
+
+	public void start() {
+		ChannelInitializer<SocketChannel> initializer = new ChannelInitializer<SocketChannel>() {
+
+			@Override
+			protected void initChannel(SocketChannel ch) {
+
+				// SSL should be the first handler in the pipeline
+				if (sslEngine != null) {
+					ch.pipeline().addLast("ssl", new SslHandler(sslEngine));
+				}
+
+				ch.pipeline()
+					.addLast(new HttpClientCodec())
+					.addLast(new HttpObjectAggregator(1024 * 1024))
+					.addLast(handler)
+					.addLast(new PipelineErrorHandler(LOG));
+			}
+		};
+
+		NioEventLoopGroup group = new NioEventLoopGroup(1);
+
+		bootstrap = new Bootstrap();
+		bootstrap
+			.group(group)
+			.channel(NioSocketChannel.class)
+			.handler(initializer);
+
+		LOG.info("Rest client endpoint started.");
```

Comment:
I actually think it is a better design to do it in the constructor. That way we don't have the additional call to `start` which is necessary in the `RestServerEndpoint` case because of the `initializeHandlers` methods.